man/man1/sacdata2inv.1
.TH sacdata2inv 1 "27 April 2023" "MTINV Version 4.0.1" "MTINV Toolkit"

.SH NAME 
sacdata2inv \- processes SAC data files for moment tensor inversion by mtinv

.SH SYNOPSIS
sacdata2inv par=(string) path=(string) respdir=(string) [no]dumpsac [no]verbose [no]parallel

.SH DESCRIPTION
sacdata2inv takes input from command line and the parameter file mtinv.par (same PAR file used by glib2inv, and mtinv)
about the station information, window processing, filtering.
The path parameter points to the directory path where the 3-component SAC files are located.
The application generates input files for moment tensor inversion using "mtinv" *.data files one per station.network.location.  
The directories of the SAC files
and Sac Pole and Zero response correction files are scanned automatically.  The data files
are processed, detrended, bandpass filtered, interpolated, windowed, scaled to the correct units,
tapered and rotated into the Vertical, Radial and Transverse (Z,R,T) coordinate system.  
Writes out *.data files for mtinv.
The stations are specified in mtinv.par (see below for format).

.nr step 1 1
Processing steps (data and Green's functions processed identically):
.IP \n[step] 3
scale amplitudes after converting meters to centimeters
.IP \n+[step]
remove trend
.IP \n+[step]
window in time
.IP \n+[step]
correct instrument response (counts -> volts -> meters displacement)
.IP \n+[step]
remove the amplitude of first sample (avoids truncated step function artificat)
.IP \n+[step]
bandpass filter
.IP \n+[step]
differentiate from displacement to velocity (if needed)
.IP \n+[step]
convolve source-time function (triangle or trapezoid; default is do nothing--delta function)
.IP \n+[step]
bandpass filter
.IP \n+[step]
interpolate using fft to new sampling rate
.IP \n+[step]
taper, only taper data since synthetics are not windowed
.IP \n+[step]
rotate channels 1 and 2 horizontals to radial and transverse or tangential
.IP \n+[step]
compute and store signal-to-noise ratios for each channel using signal and pre-event windows

.SH New Features:
(1) Rotational Green's functions
(2) Mij formated Green's functions

.SH REQUIRED PARAMETERS

.TP
.B par={string}
The file name of the mtinv.par parameter file.  See format of this file below (PARAMETER FILE FORMAT).

.TP
.B path={string}
The relative or absolute path to where the SAC files are stored. Use rdseed -d -p -f myrdseed.seed to
extract SAC and SAC pole zero files from a seed file. Since rdseed is deprecated, then use unpack.csh. 

.TP
.B respdir={string}
The relative or absolute path to where the SAC pole zero files are stored.  Use rdseed -d -p -f myrdseed.seed to
extract SAC and SAC pole zero files from a seed file. Since rdseed is deprecated, then use unpack.csh.

.SH OPTIONAL PARAMETERS

.TP
.B [no]verbose
Verbosy output for debugging is \fIverbose\fP and for no verbosy output is \fInoverbose\fP [optional, default no verbose].

.TP
.B [no]dumpsac
Write out "Dump" the processed data as SAC formatted binary files [optional, default nodump].

.TP
.B [no]parallel
The parallel option performs processing of each station in separate thread
simultaneously (up to 50) otherwise serial operation processes each station at a time
(default off).

.SH PARAMETER FILE FORMAT (usually named mtinv.par)

.TP
.B #
A '#' in the first column signifies that the following is a non printing comment. Also used to remove station from mtinv. Use (y or n) column 15 to predict instead of removing a station.

.TP
.B CM
A 'CM ' in columns 1-3 marks the beginning of a comment, typically the region, area, city and country name
for reference purposes only.

.TP
.B OT
A 'OT ' in columns 1-3 is followed by the earthquake origin time string in year/month/day,hour:minute:seconds format, (e.g.,
2005/06/12,15:41:46.000 or 2005-06-12T15:41:46.000).  Origin time is required for setting the origin time in the
synthetics and shifting the observed seismograms in time to improve the fit the synthetics and percent double couple component.

.TP
.B  EV
A 'EV ' in columns 1-3 is followed by six free formatted floating point values:  strike, dip, rake, Mw, event_longitude, event_latitude, and event_depth.  Only event_longitude and event_latitude need to be set for the correct calculation of the source and receiver azimuth and distance.  Should be the same as the \fIevla=\fP and \fIevlo=\fP values used to calculate the Green's function library using "mkgrnlib".  Other values can be set to zero and are only used to forward calculate synthetic ground displacements when the dump option is set (dumpsac).

.TP
.B MT (OPTIONAL - EXPERIMENTAL)
A 'MT' in columns 1-3 is followed by seven free formatted floating point values: Mxx, Mxy, Mxz, Myy, Myz, Mzz, Mo.
The normalized moment tensor elements and seismic moment in dyne*cm.

.TP
All other lines are for station and synthetic data processing parameters.  There are 18 columns in the following order:
.sp
.B sta, net, loc, model, np, pas, lf, hf, nt, dt, tr, tt, grdmo, mulfac, used, ts, weight, wavetype, comments

.B sta
station code (see stadb file).

.B net
network code (see stadb file).

.B loc
location code (empty quotes for null "" otherwise 00, 10, 01, 02, 60, etc...)

.B model
velocity model name (without .mod extension)

.B np
number of poles (0,1,2,3,4) for Butterworth bandpass filter

.B pas
number of passes (1 or 2) for Butterworth bandpass filter

.B lf
low frequency corner in Hz for Butterworth bandpass filter

.B hf
high frequency corner in Hz for Butterworth bandpass filter

.B nt
number of points (in power of 2)

.B dt
sampling rate in sec/sample

.B tr
rise time or duration of ramp in trapezoid function in seconds

.B tt
duration of boxcar portion of trapezoid function in seconds (tt=0 then triangle function of duration 2*tr)

.B grdmo(v or d)
Ground motion type is either 'd' for displacement or 'v' for velocity (no default)

.B mulfac
Multiplcation factor applied to all components for this station. Useful for applying gain corrections easily.
Default is 1.

.B used (y or n)
Use this station for inversion 'y' or just make a preduction 'n'

.B ts
Time shift in sec for all components in seconds.  Negative is backward time shift in time and positive shifts are forward shift in time.
Default is 0;

.B weight
Weight factor applied in A matrix in MT inversion

.B wavetype
Green's function: Surf/Pnl or Rotational

.B comments
Free format comments are allowed after column 18. In autogenerated file src-rec distance and azimuth are listed for reference.

.SH EXAMPLE PARAMETER FILE (mtinv.par)
.br
#### REGION COMMENT ############################
.br
CM New Madrid, MO
.br
#### Date and Origin Time ######################
.br
OT 2021-11-18T02:53:04.00
.br
#### Forward Calculations ######################
.br
##    stk    dip    rak   Mw  evlo  evla   Z ##########
.br
EV -999.0 -999.0 -999.0  0.0    -90.543    36.9077  15.0
.br
#####################################################################################
.br
#sta net loc mod np pas lf hf  nt   dt   tr  tt v/d mulfac used ts0 wt wvtyp  #
.br
CGM3 NM 00 cus 3 2 0.075 0.15 1024 0.05 0.0 0.0 d 1.0 y +0.0 1.0 Surf/Pnl # R=89 Az=61
.br
PENM NM 00 cus 3 2 0.075 0.15 1024 0.05 0.0 0.0 d 1.0 y +0.0 1.0 Surf/Pnl # R=96 Az=122
.br
HENM NM 00 cus 3 2 0.075 0.15 1024 0.05 0.0 0.0 d 1.0 y +0.0 1.0 Surf/Pnl # R=97 Az=102
.br
GNAR NM 00 cus 3 2 0.075 0.15 1024 0.05 0.0 0.0 d 1.0 y +0.0 1.0 Surf/Pnl # R=114 Az=156
.br
CCM  IU 00 cus 3 2 0.075 0.15 1024 0.07 0.0 0.0 d 1.0 y +0.0 1.0 Surf/Pnl # R=141 Az=334
.br
CCM  IU 10 cus 3 2 0.075 0.15 1024 0.07 0.0 0.0 d 1.0 y +0.0 1.0 Surf/Pnl # R=141 Az=334
.br
SIUC NM 00 cus 3 2 0.075 0.15 1024 0.07 0.0 0.0 d 1.0 y +0.0 1.0 Surf/Pnl # R=147 Az=52
.br
SLM  NM "" cus 3 2 0.075 0.15 1024 0.08 0.0 0.0 d 1.0 y +0.0 1.0 Surf/Pnl # R=193 Az=8
.br
#######################################################################################
.br
#WVT IU 00 cus 3 2 0.075 0.15 1024 0.11 0.0 0.0 d 1.0 n +0.0 +1.0 Surf/Pnl # R=257 Az=109
.br
#WVT IU 10 cus 3 2 0.075 0.15 1024 0.11 0.0 0.0 d 1.0 n +0.0 +1.0 Surf/Pnl # R=257 Az=109
.br

.SH Example C-Shell Script
.br
#!/bin/csh
.br
### PROCESS DATA ###
sacdata2inv par=mtinv.par path=../Data respdir=../Resp noverbose nodumpsac parallel
.br

.SH "SEE ALSO"
.IR glib2inv (1),
.IR setupMT (1),
.IR mtinv (1)


src/sacdata2inv.c
/*** reads SAC files and outputs *.data files processed (3-C) data for moment tensor inversion - mtinv.c ***/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <math.h>
#include <pthread.h>

#include "../include/mt.h"         /** global datatype and structure declarations **/
#include "../include/nrutil.h"     /** numerical recipes **/

char progname[128];

int main( int ac, char **av )
{

/****************************************************************************************/
/*** event information stuff including 3-component                                    ***/
/****************************************************************************************/

	EventInfo *ev;
	int ista, nsta;
        char evinfo_filename[128];

	/*** glib2inv_subs.c ***/
	EventInfo *glib2inv_get_input_parameters(
		char *filename,
		EventInfo *ev,
		int *n,
		int verbose );

/****************************************************************************************/
/*** Greens function header and synthetics library structure                          ***/
/****************************************************************************************/

	Greens **grn;
        FILE *fpg;
	int iz, nz;
	float *z;

/****************************************************************************************/
/*** sac file handaling, data processing, and data filtering parameters               ***/
/****************************************************************************************/

/** max number of files in a directory is 2048 ***/

	char SAC_DATA_pathname[256], respdir[256];

/****************************************************************************************/
/*** output options                                                                   ***/
/****************************************************************************************/
        int verbose=0;          /*** verbose   1=yes 0=no default no verbose ***/
	int dumpsac=0;          /*** dump the processed and filtered data as SAC binary files ***/
	int ienvelope = 0;     /*** envelope is deprecated gene.ichinose 07-01-2024 ***/
	int iparallel = 0;

	char currentdir[128];
	
/*******************************/
/*** subroutine prototypes   ***/
/*******************************/

	/*** sacdata2inv_subs.c ***/
	void  writesacfile( EventInfo *ev );

	FILE *fp_SNR;

/*** misc ***/
	void  Print_Usage(void);
	int   setpar(int,char **);
	int mstpar( char *, char *, void * );
        int getpar( char *, char *, void * );
	void  endpar();
	char exec_pathname[128];
	char *shorten_path( char *pathname, char *filename );

/*** sacdata2inv_loaddata.c ***/
	void sacdata2inv_loaddata(
	        EventInfo *ev,
        	int ista,
        	char *sacdatadir,
        	char *respdir,
        	char *currentdir,
        	int verbose );

/*** sacdata2inv_serial.c  ***/
	void sacdata2inv_serial( 
		EventInfo *ev, 
		int ista,
		int ienvelope,
		int dumpsac, 
		int verbose );

/**** sacdata2inv_parallel *****/
	void sacdata2inv_parallel(
        	EventInfo *ev,
        	int nsta,
        	int ienvelope,
        	int dumpsac,
        	int verbose );

/****************************************************************************************/
/*** get the directory location for output files                                      ***/
/****************************************************************************************/
	strcpy( exec_pathname, av[0] );
	shorten_path( exec_pathname, progname );

	if( verbose )
	{
	  fprintf( stdout, "%s: STDOUT: Version=%s ReleaseDate=%s exec full path=%s\n",
		progname, Version_Label, Version_Date, exec_pathname );
	}

	fprintf( stderr, "%s: STDERR: Version=%s ReleaseDate=%s exec full path=%s\n",
		progname, Version_Label, Version_Date, exec_pathname );

/****************************************************************************************/
/*** USAGE                                                                            ***/
/****************************************************************************************/

	if( ac <= 1 ) Print_Usage();

/****************************************************************************************/
/*** get the input parameters from command line                                       ***/
/****************************************************************************************/

	strcpy( SAC_DATA_pathname, ".\0" );
	strcpy( respdir, ".\0" );
        setpar(ac,av);
        mstpar( "par",     "s", &evinfo_filename );
        getpar( "path",    "s", &SAC_DATA_pathname );
	getpar( "respdir", "s", &respdir );
        getpar( "verbose", "b", &verbose );
        getpar( "dumpsac", "b", &dumpsac );
	getpar( "parallel", "b", &iparallel );
	endpar();

/****************************************************************************************/
/****************************************************************************************/
	getcwd( currentdir, 128 );
/*
	if( verbose )
	{
	  fprintf( stdout, "\n\n%s: current working directory is %s\n", 
		progname, currentdir );
	}
*/

/**************************************************************/
/*** get the input parameters from the input parameter file ***/
/**************************************************************/
	ev  = (EventInfo *)malloc(sizeof(EventInfo));
	ev  = (EventInfo *)
	  glib2inv_get_input_parameters( 
		evinfo_filename, 
		ev, 
		&nsta, 
		verbose );

/*********************************************************************/
/*** loop over each station and load the Green's function header   ***/
/*** to load in the values for Reduction Velocity and ts0          ***/
/*********************************************************************/

	grn = (Greens **) malloc(nsta*sizeof(Greens *));
	for( ista=0; ista<nsta; ista++ )
	{
		if(verbose)
		{
		  fprintf(stdout, "%s: loading green's functions %s\n", 
			progname, ev[ista].glib_filename );
		}

		if( (fpg=fopen( ev[ista].glib_filename, "rb" )) == NULL )
		{
			fprintf( stderr, "%s: Fatal Error, cannot open file %s\n",
				progname, ev[ista].glib_filename );
			exit(-1);
		}

		fread( &nz, sizeof(int), 1, fpg );
		z = malloc( nz*sizeof(float) );
		fread( &z[0], nz*sizeof(float), 1, fpg );
		grn[ista] = (Greens *)malloc(nz*sizeof(Greens) );

		for( iz=0; iz<nz; iz++ )
		{
			fread( &grn[ista][iz], sizeof(Greens), 1, fpg );
			if(verbose)
			{
			  fprintf( stdout,
			    "%s: iz=%d z=%g rdist=%g az=%g evdp=%g t0=%g dt=%g nt=%d %s\n",
				progname,
				iz,
				z[iz],
				grn[ista][iz].rdist,
				grn[ista][iz].az,
				grn[ista][iz].evdp,
				grn[ista][iz].t0,
				grn[ista][iz].dt,
				grn[ista][iz].nt,
				grn[ista][iz].filename );
			}
		}
		fclose(fpg);

		ev[ista].redv   = grn[ista][0].redv;
		ev[ista].ts0    = grn[ista][0].ts0;
		ev[ista].tstart = grn[ista][0].tstart;
		ev[ista].tend   = grn[ista][0].tend;
		ev[ista].rdist  = grn[ista][0].rdist;
	}

/********************************************************************/
/*** loop over each station and load its SAC files for all 3 comp ***/
/********************************************************************/
	fprintf( stdout, "%s: %d number of stations to process.\n",
		progname, nsta );

	sacdata2inv_loaddata( ev, nsta, SAC_DATA_pathname, respdir, currentdir, verbose );

	if( iparallel )
	{
		sacdata2inv_parallel(
			ev,
			nsta, 
			ienvelope,
			dumpsac,
			verbose );
	}
	else
	{
	  for( ista=0; ista<nsta; ista++ )
	  {
		sacdata2inv_serial(
			&(ev[ista]),
			ista, 
			ienvelope,
			dumpsac,
			verbose );
	  }
	}

/********************************************************************/
/*** output snr                                                   ***/
/********************************************************************/

	fprintf( stdout, "\n\n%s: Writting to file snr.out\n", progname );
	fp_SNR = fopen( "snr.out", "w" );

	for( ista=0; ista<nsta; ista++ )
	{
		if( strcmp( ev[ista].loc, "" ) == 0 ) strcpy( ev[ista].loc, "--" );

		fprintf( stdout, 
"\n%s.%s.%s\n\tZ:  Tnoi=%.1f Tsig=%.1f log10amp=%.5f/%.5f snr=%.5f\n\tNS: Tnoi=%.1f Tsig=%.1f log10amp=%.5f/%.5f snr=%.5f\n\tEW: Tnoi=%.1f Tsig=%.1f log10amp=%.5f/%.5f snr=%.5f\n",
	ev[ista].net, 
	ev[ista].stnm, 
	ev[ista].loc,
  ev[ista].z.pha[NOISE].period,  ev[ista].z.pha[SIGNAL].period,  log10(ev[ista].z.pha[SIGNAL].amp),  log10(ev[ista].z.pha[NOISE].amp),  ev[ista].z.P2P_snr, 
  ev[ista].ns.pha[NOISE].period, ev[ista].ns.pha[SIGNAL].period, log10(ev[ista].ns.pha[SIGNAL].amp), log10(ev[ista].ns.pha[NOISE].amp), ev[ista].ns.P2P_snr,
  ev[ista].ew.pha[NOISE].period, ev[ista].ew.pha[SIGNAL].period, log10(ev[ista].ew.pha[SIGNAL].amp), log10(ev[ista].ew.pha[NOISE].amp), ev[ista].ew.P2P_snr );

		fprintf( fp_SNR, 
		  "%8s %8s %8s %8s %6.3f %6.3f %7.1f %3.0f %3.0f %7.2f %7.2f %9.5f %9.5f %11.5f\n",
			ev[ista].stnm,
			ev[ista].net,
			ev[ista].loc,
			ev[ista].z.s.kcmpnm,
			ev[ista].lf,
			ev[ista].hf, 
			ev[ista].rdistkm,
			ev[ista].az,
			ev[ista].baz,
			ev[ista].z.pha[NOISE].period,
			ev[ista].z.pha[SIGNAL].period,
			log10(ev[ista].z.pha[SIGNAL].amp),
			log10(ev[ista].z.pha[NOISE].amp),
			ev[ista].z.P2P_snr );

		fprintf( fp_SNR,
                  "%8s %8s %8s %8s %6.3f %6.3f %7.1f %3.0f %3.0f %7.2f %7.2f %9.5f %9.5f %11.5f\n",
                        ev[ista].stnm,
                        ev[ista].net,
			ev[ista].loc,
                        ev[ista].ns.s.kcmpnm,
			ev[ista].lf,
                        ev[ista].hf,
                        ev[ista].rdistkm,
                        ev[ista].az,
                        ev[ista].baz,
			ev[ista].ns.pha[NOISE].period,
			ev[ista].ns.pha[SIGNAL].period,
			log10(ev[ista].ns.pha[SIGNAL].amp),
			log10(ev[ista].ns.pha[NOISE].amp),
			ev[ista].ns.P2P_snr );

		fprintf( fp_SNR,
                  "%8s %8s %8s %8s %6.3f %6.3f %7.1f %3.0f %3.0f %7.2f %7.2f %9.5f %9.5f %11.5f\n",
			ev[ista].stnm,
                        ev[ista].net,
			ev[ista].loc,
			ev[ista].ew.s.kcmpnm,
			ev[ista].lf,
                        ev[ista].hf,
                        ev[ista].rdistkm,
                        ev[ista].az,
                        ev[ista].baz,
			ev[ista].ew.pha[NOISE].period, 
			ev[ista].ew.pha[SIGNAL].period, 
			log10(ev[ista].ew.pha[SIGNAL].amp), 
			log10(ev[ista].ew.pha[NOISE].amp), 
			ev[ista].ew.P2P_snr );

	} /*** loop over ista ***/
	fclose(fp_SNR);

	if(verbose)
	fprintf(stdout, "\n%s: sacdata2inv(): STDOUT program finished. Bye-Bye!\n\n\n", progname );
	fprintf(stderr, "\n%s: sacdata2inv(): STDERR program finished. Bye-Bye!\n\n\n", progname );

	return 0;

} /*** end of main ***/

void Print_Usage( void )
{
	fprintf( stderr, "\nUsage: %s par= path= respdir= [no]dumpsac [no]verbose [no]parallel\n",
		progname );

	fprintf( stderr, "\nREQUIRED PARAMETERS:\n" );

	fprintf( stderr, "\t par=mtinv.par        parameter file\n" );

	fprintf( stderr, "\t path=/my/data/dir    directory path for *.SAC data files\n" );
	fprintf( stderr, "\t\t    DEFAULT is current working directory\n" );

	fprintf( stderr, "\t respdir=/my/resp/dir directory path for SAC POLE ZERO Response files\n" );
	fprintf( stderr, "\t\t    DEFAULT is current working directory\n" );

	fprintf( stderr, "\nOPTIONAL PARAMETERS:\n" );
	fprintf( stderr, "\t [no]verbose         be verbosy DEFAULT is off\n" );
	fprintf( stderr, "\t [no]dumpsac         write out the data as SAC formatted files DEFAULT off\n" );
	fprintf( stderr, "\t [no]parallel        use multithreading, each station runs on seperate thread DEFAULT off\n\n" );

	fprintf( stderr, "\t see file snr.out for Signal-to-noise ratios and period for peak-to-peak amplitudes\n\n" );
}
